test/functional/p2p_invalid_messages.py

```diff
@@ -248,6 +250,34 @@ def test_oversized_headers_msg(self):
         size = MAX_HEADERS_RESULTS + 1
         self.test_oversized_msg(msg_headers([CBlockHeader()] * size), size)
 
+    def test_invalid_pow_headers_msg(self):
+        self.log.info("Test headers message with invalid proof-of-work is logged as misbehaving and disconnects peer")
+        blockheader_tip_hash = self.nodes[0].getbestblockhash()
+        blockheader_tip = from_hex(CBlockHeader(), self.nodes[0].getblockheader(blockheader_tip_hash, False))
+
+        # send valid headers message first
+        assert_equal(self.nodes[0].getblockchaininfo()['headers'], 0)
+        blockheader = CBlockHeader()
+        blockheader.hashPrevBlock = int(blockheader_tip_hash, 16)
+        blockheader.nTime = int(time.time())
+        blockheader.nBits = blockheader_tip.nBits
+        blockheader.rehash()
+        while not blockheader.hash.startswith('0'):
+            blockheader.nNonce += 1
+            blockheader.rehash()
+        good_peer = self.nodes[0].add_p2p_connection(P2PInterface())
+        good_peer.send_and_ping(msg_headers([blockheader]))
+        assert_equal(self.nodes[0].getblockchaininfo()['headers'], 1)
+
+        # invalidate PoW
+        while not blockheader.hash.startswith('f'):
+            blockheader.nNonce += 1
+            blockheader.rehash()
+        bad_peer = self.nodes[0].add_p2p_connection(P2PInterface())
```

Comment:
Do we need to create a new peer here (`bad_peer`)? Couldn't we use the same used before? e.g.:
```diff
diff --git a/test/functional/p2p_invalid_messages.py b/test/functional/p2p_invalid_messages.py
index 7c6777759..39505c115 100755
--- a/test/functional/p2p_invalid_messages.py
+++ b/test/functional/p2p_invalid_messages.py
@@ -265,18 +265,17 @@ class InvalidMessagesTest(BitcoinTestFramework):
         while not blockheader.hash.startswith('0'):
             blockheader.nNonce += 1
             blockheader.rehash()
-        good_peer = self.nodes[0].add_p2p_connection(P2PInterface())
-        good_peer.send_and_ping(msg_headers([blockheader]))
+        peer = self.nodes[0].add_p2p_connection(P2PInterface())
+        peer.send_and_ping(msg_headers([blockheader]))
         assert_equal(self.nodes[0].getblockchaininfo()['headers'], 1)
 
         # invalidate PoW
         while not blockheader.hash.startswith('f'):
             blockheader.nNonce += 1
             blockheader.rehash()
-        bad_peer = self.nodes[0].add_p2p_connection(P2PInterface())
         with self.nodes[0].assert_debug_log(['Misbehaving', 'header with invalid proof of work']):
-            bad_peer.send_message(msg_headers([blockheader]))
-            bad_peer.wait_for_disconnect()
+            peer.send_message(msg_headers([blockheader]))
+            peer.wait_for_disconnect()
```

Reply:
Good point, done.